### PR TITLE
feat: headers, utils, and REST client construction

### PR DESCRIPTION
**Pull request by software-dov**
GAPICs that build REST clients now include factory functions that
create REST clients, as well as commensurate utilities.

### CODE REVIEW

File: internal/gengapic/genrest.go
Comment (vchudnov-g) [inline internal/gengapic/genrest.go:92]: Do we need an import for `internaloption`? I see one in gengrpc.go, but not in genrest.go.
(Also, I seem to be missing this in the `*.want` golden files.)

Comment (vchudnov-g) [inline internal/gengapic/genrest.go:92]: Do library users have the ability to manually set the protocol to be HTTP rather than HTTPS at client instantiation time? If not, until we do have such an ability, would this be the right place to automatically set it to be HTTP if the endpoint starts with `localhost`? (This would be a stop-gap measure until we can have users do it at client creation time; I can point you to the internal doc on the topic. Or see the Python PR....)

This would be needed to hook up the Showcase integration tests; we can make those two changes at the same time.

File: internal/gengapic/genrest.go
Comment (vchudnov-g) [inline internal/gengapic/genrest.go:138]: Maybe use `0.0` rather than `UNKNOWN` to make it clear this is an explicit choice (that way our data pipelines can fill in UNKNOWN for missing values, say)

File: internal/gengapic/genrest.go
Comment (vchudnov-g) [inline internal/gengapic/genrest.go:397]: nit: Suggest making the field name "canonical" as per the http package doc: Capitalized-Kebab-Case. Below as well.

**Review by vchudnov-g — COMMENTED** [review]
lgtm, but have @noahdietz or @codyoss TAL as well.

File: internal/gengapic/genrest.go
Comment (codyoss) [reply internal/gengapic/genrest.go:138]: Lets keep this as UNKNOWN for now. We have other automation that post processes this value into something more meaningful

File: internal/gengapic/genrest.go
Comment (codyoss) [reply internal/gengapic/genrest.go:92]: At least in apiary land the default endpoints contain the scheme as well. So if a user wants non-https they just specify when they override it.

Example : 
- `internaloption.WithDefaultEndpoint(https://secretmanager.googleapis.com/)`
- ``internaloption.WithDefaultMTLSEndpoint(https://secretmanager.mtls.googleapis.com/)

File: internal/gengapic/testdata/empty_client_init.want
Comment (codyoss) [inline internal/gengapic/testdata/empty_client_init.want:173]: nit: docs

File: internal/gengapic/genrest.go
Comment (vchudnov-g) [reply internal/gengapic/genrest.go:92]: So this is something we'll be able to do immediately for REST clients at instantiation time, once this PR is merged? Perfect.

File: internal/gengapic/genrest.go
Comment (vchudnov-g) [reply internal/gengapic/genrest.go:138]: This post-processing happens before publication? OK. Like I say, once the library is in use I don't want to pollute the space of values we provide.

Comment (codyoss) [reply internal/gengapic/genrest.go:138]: Yup, it happens before the PR is raised to land the changes: https://github.com/googleapis/google-cloud-go/blob/master/internal/gapicgen/generator/gapics.go#L85-L97

**Review by codyoss — COMMENTED** [review]
LGTM, I defer to your team for sign-off.

**Review by vchudnov-g — APPROVED** [review]

**Pull request merged**
